docs/rfcs/optional-config-type.md

```diff
@@ -0,0 +1,276 @@
+not a valid configuration value and can be taken to mean that the option is
+disabled, but in certain cases it can indicate a default value, necessitating a
+way to represent the presence of a value without using a valid value for the
+type.
+
+Using standard Go without inventing any new types, the two most straightforward
+ways to accomplish this are:
+
+1. Using a separate boolean field to indicate whether the field is enabled or
+   disabled.
+2. Making the type a pointer, which makes a `nil` pointer represent that a
+   value is not present, and a valid pointer represent the desired value.
+
+Each of these approaches has deficiencies: Using a separate boolean field
+requires the user to set the boolean field to `true` in addition to setting the
+actual config option, leading to suboptimal UX. Using a pointer value has a few
+drawbacks:
+
+1. It may not be immediately obvious to a new user that a pointer type indicates
+   a field is optional.
+2. The distinction between values that are conventionally pointers (e.g. gRPC
+   configs) and optional values is lost.
+3. Setting a default value for a pointer field when decoding will set the field
+   on the resulting config struct, and additional logic must be done to unset
+   the default if the user has not specified a value.
+4. The potential for null pointer exceptions is created.
+5. Config structs are generally intended to be immutable and may be passed
+   around a lot, which makes the mutability property of pointer fields
+   an undesirable property.
+
+## Optional types
+
+Go does not include any form of Optional type in the standard library, but other
+popular languages like Rust and Java do. We could implement something similar in
+our config packages that allows us to address the downsides of using pointers
+for optional config fields.
+
+## Basic definition
+
+A production-grade implementation will not be discussed or shown here, but the
+basic implementation for an Optional type in Go could look something like:
+
+```golang
+type Optional[T any] struct {
+	hasValue bool
+	value    T
+
+	defaultVal T
+}
+
+func Some[T any](value T) Optional[T] {
+	return Optional[T]{value: value, hasValue: true}
+}
+
+func None[T any](value T) Optional[T] {
+	return Optional[T]{}
+}
+
+func WithDefault[T any](defaultVal T) Optional[T] {
+	return Optional[T]{defaultVal: defaultVal}
+}
+
+func (o Optional[T]) HasValue() bool {
+	return o.hasValue
+}
+
+func (o Optional[T]) Value() T {
+    return o.value
+}
+
+func (o Optional[T]) Default() T {
+    return o.defaultVal
+}
+```
+
+## Use cases
+
+Optional types can fulfill the following use cases we have when decoding config.
+
+### Representing optional config fields
+
+To use the optional type to mark a config field as optional, the type can simply
+be used as a type parameter to `Optional[T]`. The following config struct shows
+how this may look, both in definition and in usage:
+
+```golang
```

Comment:
More than golang code, would be good to show how users in yaml will use it.

Reply:
there's no expected impact on YAML

Reply:
@bogdandrutu does Yuri's comment solve your question?

Reply:
Clarified inline on 5d5f6527f1d4444c657d30c19f3055e781830913